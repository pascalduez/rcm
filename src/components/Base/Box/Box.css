// -------------------- Box model ----------------------------------------------

.borderBox {
  box-sizing: border-box;
}

.contentBox {
  box-sizing: content-box;
}

.paddingBox {
  box-sizing: padding-box;
}

// -------------------- Reset --------------------------------------------------

// /!\ Should not be used for overriding other composable classes /!\
// Only browser defaults or vendors styles.
// That's why we declare them before the other spacing props in the cascade.

.marginNone {
  margin: 0;
}

.marginTopNone {
  margin: 0;
}

.marginRightNone {
  margin: 0;
}

.marginBottomNone {
  margin: 0;
}

.marginLeftNone {
  margin: 0;
}

.marginXNone {
  margin: 0;
}

.marginYNone {
  margin: 0;
}

.paddingNone {
  padding: 0;
}

.paddingTopNone {
  padding: 0;
}

.paddingRightNone {
  padding: 0;
}

.paddingBottomNone {
  padding: 0;
}

.paddingLeftNone {
  padding: 0;
}

.paddingXNone {
  padding: 0;
}

.paddingYNone {
  padding: 0;
}

// -------------------- Spacing ------------------------------------------------

.marginXsmall {
  margin: $spacing-xsmall-mobile;
}

.marginSmall {
  margin: $spacing-small-mobile;
}

.marginBase {
  margin: $spacing-mobile;
}

.marginMedium {
  margin: $spacing-medium-mobile;
}

.marginLarge {
  margin: $spacing-large-mobile;
}

.marginTopXsmall {
  margin-top: $spacing-xsmall-mobile;
}

.marginTopSmall {
  margin-top: $spacing-small-mobile;
}

.marginTopBase {
  margin-top: $spacing-mobile;
}

.marginTopMedium {
  margin-top: $spacing-medium-mobile;
}

.marginTopLarge {
  margin-top: $spacing-large-mobile;
}

.marginRightXsmall {
  margin-right: $spacing-xsmall-mobile;
}

.marginRightSmall {
  margin-right: $spacing-small-mobile;
}

.marginRightBase {
  margin-right: $spacing-mobile;
}

.marginRightMedium {
  margin-right: $spacing-medium-mobile;
}

.marginRightLarge {
  margin-right: $spacing-large-mobile;
}

.marginBottomXsmall {
  margin-bottom: $spacing-xsmall-mobile;
}

.marginBottomSmall {
  margin-bottom: $spacing-small-mobile;
}

.marginBottomBase {
  margin-bottom: $spacing-mobile;
}

.marginBottomMedium {
  margin-bottom: $spacing-medium-mobile;
}

.marginBottomLarge {
  margin-bottom: $spacing-large-mobile;
}

.marginLeftXsmall {
  margin-left: $spacing-xsmall-mobile;
}

.marginLeftSmall {
  margin-left: $spacing-small-mobile;
}

.marginLeftBase {
  margin-left: $spacing-mobile;
}

.marginLeftMedium {
  margin-left: $spacing-medium-mobile;
}

.marginLeftLarge {
  margin-left: $spacing-large-mobile;
}

.paddingXsmall {
  padding: $spacing-xsmall-mobile;
}

.paddingSmall {
  padding: $spacing-small-mobile;
}

.paddingBase {
  padding: $spacing-mobile;
}

.paddingMedium {
  padding: $spacing-medium-mobile;
}

.paddingLarge {
  padding: $spacing-large-mobile;
}

.paddingTopXsmall {
  padding-top: $spacing-xsmall-mobile;
}

.paddingTopSmall {
  padding-top: $spacing-small-mobile;
}

.paddingTopBase {
  padding-top: $spacing-mobile;
}

.paddingTopMedium {
  padding-top: $spacing-medium-mobile;
}

.paddingTopLarge {
  padding-top: $spacing-large-mobile;
}

.paddingRightXsmall {
  padding-right: $spacing-xsmall-mobile;
}

.paddingRightSmall {
  padding-right: $spacing-small-mobile;
}

.paddingRightBase {
  padding-right: $spacing-mobile;
}

.paddingRightMedium {
  padding-right: $spacing-medium-mobile;
}

.paddingRightLarge {
  padding-right: $spacing-large-mobile;
}

.paddingBottomXsmall {
  padding-bottom: $spacing-xsmall-mobile;
}

.paddingBottomSmall {
  padding-bottom: $spacing-small-mobile;
}

.paddingBottomBase {
  padding-bottom: $spacing-mobile;
}

.paddingBottomMedium {
  padding-bottom: $spacing-medium-mobile;
}

.paddingBottomLarge {
  padding-bottom: $spacing-large-mobile;
}

.paddingLeftXsmall {
  padding-left: $spacing-xsmall-mobile;
}

.paddingLeftSmall {
  padding-left: $spacing-small-mobile;
}

.paddingLeftBase {
  padding-left: $spacing-mobile;
}

.paddingLeftMedium {
  padding-left: $spacing-medium-mobile;
}

.paddingLeftLarge {
  padding-left: $spacing-large-mobile;
}

@media screen and (min-width: $screen-small) {
  .marginXsmall {
    margin: $spacing-xsmall;
  }

  .marginSmall {
    margin: $spacing-small;
  }

  .marginBase {
    margin: $spacing;
  }

  .marginMedium {
    margin: $spacing-medium;
  }

  .marginLarge {
    margin: $spacing-large;
  }

  .marginTopXsmall {
    margin-top: $spacing-xsmall;
  }

  .marginTopSmall {
    margin-top: $spacing-small;
  }

  .marginTopBase {
    margin-top: $spacing;
  }

  .marginTopMedium {
    margin-top: $spacing-medium;
  }

  .marginTopLarge {
    margin-top: $spacing-large;
  }

  .marginRightXsmall {
    margin-right: $spacing-xsmall;
  }

  .marginRightSmall {
    margin-right: $spacing-small;
  }

  .marginRightBase {
    margin-right: $spacing;
  }

  .marginRightMedium {
    margin-right: $spacing-medium;
  }

  .marginRightLarge {
    margin-right: $spacing-large;
  }

  .marginBottomXsmall {
    margin-bottom: $spacing-xsmall;
  }

  .marginBottomSmall {
    margin-bottom: $spacing-small;
  }

  .marginBottomBase {
    margin-bottom: $spacing;
  }

  .marginBottomMedium {
    margin-bottom: $spacing-medium;
  }

  .marginBottomLarge {
    margin-bottom: $spacing-large;
  }

  .marginLeftXsmall {
    margin-left: $spacing-xsmall;
  }

  .marginLeftSmall {
    margin-left: $spacing-small;
  }

  .marginLeftBase {
    margin-left: $spacing;
  }

  .marginLeftMedium {
    margin-left: $spacing-medium;
  }

  .marginLeftLarge {
    margin-left: $spacing-large;
  }

  .paddingXsmall {
    padding: $spacing-xsmall;
  }

  .paddingSmall {
    padding: $spacing-small;
  }

  .paddingBase {
    padding: $spacing;
  }

  .paddingMedium {
    padding: $spacing-medium;
  }

  .paddingLarge {
    padding: $spacing-large;
  }

  .paddingTopXsmall {
    padding-top: $spacing-xsmall;
  }

  .paddingTopSmall {
    padding-top: $spacing-small;
  }

  .paddingTopBase {
    padding-top: $spacing;
  }

  .paddingTopMedium {
    padding-top: $spacing-medium;
  }

  .paddingTopLarge {
    padding-top: $spacing-large;
  }

  .paddingRightXsmall {
    padding-right: $spacing-xsmall;
  }

  .paddingRightSmall {
    padding-right: $spacing-small;
  }

  .paddingRightBase {
    padding-right: $spacing;
  }

  .paddingRightMedium {
    padding-right: $spacing-medium;
  }

  .paddingRightLarge {
    padding-right: $spacing-large;
  }

  .paddingBottomXsmall {
    padding-bottom: $spacing-xsmall;
  }

  .paddingBottomSmall {
    padding-bottom: $spacing-small;
  }

  .paddingBottomBase {
    padding-bottom: $spacing;
  }

  .paddingBottomMedium {
    padding-bottom: $spacing-medium;
  }

  .paddingBottomLarge {
    padding-bottom: $spacing-large;
  }

  .paddingLeftXsmall {
    padding-left: $spacing-xsmall;
  }

  .paddingLeftSmall {
    padding-left: $spacing-small;
  }

  .paddingLeftBase {
    padding-left: $spacing;
  }

  .paddingLeftMedium {
    padding-left: $spacing-medium;
  }

  .paddingLeftLarge {
    padding-left: $spacing-large;
  }
}

// Horizontal and vertical axis shorthands.

.marginXXsmall {
  composes: marginLeftXsmall marginRightXsmall;
}

.marginXSmall {
  composes: marginLeftSmall marginRightSmall;
}

.marginXBase {
  composes: marginLeftBase marginRightBase;
}

.marginXMedium {
  composes: marginLeftMedium marginRightMedium;
}

.marginXLarge {
  composes: marginLeftLarge marginRightLarge;
}

.marginYXsmall {
  composes: marginTopXsmall marginBottomXsmall;
}

.marginYSmall {
  composes: marginTopSmall marginBottomSmall;
}

.marginYBase {
  composes: marginTopBase marginBottomBase;
}

.marginYMedium {
  composes: marginTopMedium marginBottomMedium;
}

.marginYLarge {
  composes: marginTopLarge marginBottomLarge;
}

.paddingXXsmall {
  composes: paddingLeftXsmall paddingRightXsmall;
}

.paddingXSmall {
  composes: paddingLeftSmall paddingRightSmall;
}

.paddingXBase {
  composes: paddingLeftBase paddingRightBase;
}

.paddingXMedium {
  composes: paddingLeftMedium paddingRightMedium;
}

.paddingXLarge {
  composes: paddingLeftLarge paddingRightLarge;
}

.paddingYXsmall {
  composes: paddingTopXsmall paddingBottomXsmall;
}

.paddingYSmall {
  composes: paddingTopSmall paddingBottomSmall;
}

.paddingYBase {
  composes: paddingTopBase paddingBottomBase;
}

.paddingYMedium {
  composes: paddingTopMedium paddingBottomMedium;
}

.paddingYLarge {
  composes: paddingTopLarge paddingBottomLarge;
}

// -------------------- Place shorthands ---------------------------------------

// See https://drafts.csswg.org/css-align-3/#place-items-property
// See https://github.com/malyw/postcss-place-shorthands

// 1. Fallback.
.placeItemsCenter {
  align-items: center; // 1
  justify-items: center; // 1
  place-items: center;
}

// -------------------- Positioning schemes ------------------------------------

.positionStatic {
  position: static;
}

.positionRelative {
  position: relative;
}

.positionAbsolute {
  position: absolute;
}

.positionFixed {
  position: fixed;
}

.positionSticky {
  position: sticky;
}

// -------------------- Backgrounds --------------------------------------------

.backgroundColorWhite {
  background-color: white;
}

.backgroundColorGreyLighter {
  background-color: $color-grey-lighter;
}

.backgroundColorGreyLight {
  background-color: $color-grey-light;
}

.backgroundColorGrey {
  background-color: $color-grey;
}

.backgroundColorGreyDark {
  background-color: $color-grey-dark;
}

.backgroundColorGreyDarker {
  background-color: $color-grey-darker;
}

.backgroundColorTransparent {
  background-color: transparent;
}

// -------------------- Backgrounds: hover -------------------------------------

.hoverBackgroundColorWhite {
  &:hover {
    background-color: white;
  }
}

.hoverBackgroundColorSuccess {
  &:hover {
    background-color: $color-success;
  }
}

.hoverBackgroundColorSuccessDark {
  &:hover {
    background-color: mix(black, $color-success, 10%);
  }
}

// -------------------- Borders ------------------------------------------------

.borderColorWhite {
  border-color: white;
}

.borderColorInfo {
  border-color: $color-info;
}

.borderColorSuccess {
  border-color: $color-success;
}

.borderColorWarning {
  border-color: $color-warning;
}

.borderColorDanger {
  border-color: $color-danger;
}

.borderColorGrey {
  border-color: $color-grey;
}

.borderColorGreyDark {
  border-color: $color-grey-dark;
}

.borderColorGreyDarker {
  border-color: $color-grey-darker;
}

.borderColorTransparent {
  border-color: transparent;
}

// -------------------- Borders: hover -----------------------------------------

.hoverBorderColorWhite {
  &:hover {
    border-color: white;
  }
}

.hoverBorderColorInfo {
  &:hover {
    border-color: $color-info;
  }
}

.hoverBorderColorSuccess {
  &:hover {
    border-color: $color-success;
  }
}

.hoverBorderColorWarning {
  &:hover {
    border-color: $color-warning;
  }
}

.hoverBorderColorDanger {
  &:hover {
    border-color: $color-danger;
  }
}

.hoverBorderColorGrey {
  &:hover {
    border-color: $color-grey;
  }
}

.hoverBorderColorGreyDark {
  &:hover {
    border-color: $color-grey-dark;
  }
}

.hoverBorderColorGreyDarker {
  &:hover {
    border-color: $color-grey-darker;
  }
}

.hoverBorderColorTransparent {
  &:hover {
    border-color: transparent;
  }
}

// -------------------- Hoverable ----------------------------------------------

.hoverable {
  transition: color 0.2s, background-color 0.2s, border-color 0.2s,
    box-shadow 0.2s;

  &:hover {
    text-decoration: none;
  }
}

.interactive {
  composes: hoverable;

  &:hover {
    cursor: pointer;
  }
}

// -------------------- Separators ---------------------------------------------

.separatorTop {
  margin-top: 1px;
}

.separatorRight {
  margin-right: 1px;
}

.separatorBottom {
  margin-bottom: 1px;
}

.separatorLeft {
  margin-left: 1px;
}
