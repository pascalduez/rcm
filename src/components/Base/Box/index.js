// @flow

export { default as withBox } from './withBox';
export type {
  Props as withBoxProps,
  BackgroundColor as BoxBackgroundColor,
  HoverBackgroundColor as BoxHoverBackgroundColor,
  BorderColor as BoxBorderColor,
  HoverBorderColor as BoxHoverBorderColor,
} from './withBox';
export { default as Box } from './Box';
export type { Props as BoxProps } from './Box';
