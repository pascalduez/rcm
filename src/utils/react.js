// @flow

import type { Element } from 'react';

export const getElementDisplayName = (element: Element<*>): string =>
  element.type.displayName ||
  // Function name.
  element.type.name ||
  // Function without a name.
  (typeof element.type === 'function' ? 'No Display Name' : element.type);
