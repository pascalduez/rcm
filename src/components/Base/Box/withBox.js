// @flow

import * as React from 'react';
import classNames from 'classnames';
import camelCase from 'lodash/camelCase';

import styles from './Box.css';

type Spacing = 'xsmall' | 'small' | 'base' | 'medium' | 'large' | 'none';

type BoxSizing = 'border' | 'content' | 'padding';

type Position = 'static' | 'relative' | 'absolute' | 'fixed' | 'sticky';

export type BackgroundColor =
  | 'white'
  | 'grey-lighter'
  | 'grey-light'
  | 'grey'
  | 'grey-dark'
  | 'grey-darker'
  | 'transparent';

export type HoverBackgroundColor = 'white' | 'success' | 'success-dark';

export type BorderColor =
  | 'white'
  | 'info'
  | 'success'
  | 'warning'
  | 'danger'
  | 'grey'
  | 'grey-dark'
  | 'grey-darker'
  | 'transparent';

export type HoverBorderColor =
  | 'white'
  | 'success'
  | 'warning'
  | 'danger'
  | 'grey'
  | 'grey-dark'
  | 'grey-darker'
  | 'transparent';

export type Props = {
  boxSizing?: ?BoxSizing,
  margin?: ?Spacing,
  padding?: ?Spacing,
  marginTop?: ?Spacing,
  marginRight?: Spacing,
  marginBottom?: ?Spacing,
  marginLeft?: ?Spacing,
  marginX?: ?Spacing,
  marginY?: ?Spacing,
  paddingTop?: ?Spacing,
  paddingRight?: ?Spacing,
  paddingBottom?: ?Spacing,
  paddingLeft?: ?Spacing,
  paddingX?: ?Spacing,
  paddingY?: ?Spacing,
  placeItemsCenter?: boolean,
  position?: Position,
  separatorTop?: boolean,
  separatorRight?: boolean,
  separatorBottom?: boolean,
  separatorLeft?: boolean,
  backgroundColor?: ?BackgroundColor,
  hoverBackgroundColor?: ?HoverBackgroundColor,
  borderColor?: ?BorderColor,
  hoverBorderColor?: ?HoverBorderColor,
  hoverable?: boolean,
  interactive?: boolean,
  className?: string,
};

/**
 * Helper: box model styling component.
 */
export default function withBox(
  Comp: React.ComponentType<*>
): React.ComponentType<*> {
  function Box(props: Props) {
    const {
      boxSizing,
      margin,
      padding,
      marginTop,
      marginRight,
      marginBottom,
      marginLeft,
      marginX,
      marginY,
      paddingTop,
      paddingRight,
      paddingBottom,
      paddingLeft,
      paddingX,
      paddingY,
      placeItemsCenter,
      position,
      separatorTop,
      separatorRight,
      separatorBottom,
      separatorLeft,
      backgroundColor,
      hoverBackgroundColor,
      borderColor,
      hoverBorderColor,
      hoverable,
      interactive,
      className,
      ...extraProps
    } = props;

    const rootClasses = classNames(
      {
        [styles.borderBox]: boxSizing === 'border',
        [styles.contentBox]: boxSizing === 'content',
        [styles.paddingBox]: boxSizing === 'padding',
      },
      margin && styles[camelCase(`margin-${margin}`)],
      padding && styles[camelCase(`padding-${padding}`)],
      marginTop && styles[camelCase(`margin-top-${marginTop}`)],
      marginRight && styles[camelCase(`margin-right-${marginRight}`)],
      marginBottom && styles[camelCase(`margin-bottom-${marginBottom}`)],
      marginLeft && styles[camelCase(`margin-left-${marginLeft}`)],
      marginX && styles[camelCase(`margin-x-${marginX}`)],
      marginY && styles[camelCase(`margin-y-${marginY}`)],
      paddingTop && styles[camelCase(`padding-top-${paddingTop}`)],
      paddingRight && styles[camelCase(`padding-right-${paddingRight}`)],
      paddingBottom && styles[camelCase(`padding-bottom-${paddingBottom}`)],
      paddingLeft && styles[camelCase(`padding-left-${paddingLeft}`)],
      paddingX && styles[camelCase(`padding-x-${paddingX}`)],
      paddingY && styles[camelCase(`padding-y-${paddingY}`)],
      placeItemsCenter && styles.placeItemsCenter,
      position && styles[camelCase(`position-${position}`)],
      separatorTop && styles.separatorTop,
      separatorRight && styles.separatorRight,
      separatorBottom && styles.separatorBottom,
      separatorLeft && styles.separatorLeft,
      backgroundColor &&
        styles[camelCase(`background-color-${backgroundColor}`)],
      hoverBackgroundColor &&
        styles[camelCase(`hover-background-color-${hoverBackgroundColor}`)],
      borderColor && styles[camelCase(`border-color-${borderColor}`)],
      hoverBorderColor &&
        styles[camelCase(`hover-border-color-${hoverBorderColor}`)],
      hoverable && styles.hoverable,
      interactive && styles.interactive,
      className
    );

    return <Comp className={rootClasses} {...extraProps} />;
  }

  return Box;
}
