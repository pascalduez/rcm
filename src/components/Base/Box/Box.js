// @flow

import * as React from 'react';

import withBox, { type Props as InjectedProps } from './withBox';

export type Props = {
  tagName: string,
  className?: string,
  children?: React.Node,
} & InjectedProps;

function Box({ tagName = 'div', children, ...extraProps }: Props) {
  return React.createElement(tagName, extraProps, children);
}

const Typed = withBox(Box);

Box.displayName = 'Box';
Typed.displayName = 'Box';

export default Typed;
