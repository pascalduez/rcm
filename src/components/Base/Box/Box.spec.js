// @flow

import * as React from 'react';
import { render } from 'enzyme';

import Box from './Box';

describe('<Box />', () => {
  it('should render the proper default markup and classes', () => {
    const wrapper = render(<Box />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should render the proper markup following the `tagName` property', () => {
    const wrapper = render(<Box tagName="span" />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should add the proper classes based on props', () => {
    const wrapper = render(
      <Box
        boxSizing="padding"
        margin="small"
        marginLeft="medium"
        paddingTop="large"
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should pass extra props through', () => {
    const wrapper = render(<Box style={{ display: 'block' }}>Just a text</Box>);

    expect(wrapper).toMatchSnapshot();
  });
});
