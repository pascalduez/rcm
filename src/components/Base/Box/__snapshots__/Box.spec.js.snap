// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Box /> should add the proper classes based on props 1`] = `
<div
  class="undefined"
/>
`;

exports[`<Box /> should pass extra props through 1`] = `
<div
  class=""
  style="display:block"
>
  Just a text
</div>
`;

exports[`<Box /> should render the proper default markup and classes 1`] = `
<div
  class=""
/>
`;

exports[`<Box /> should render the proper markup following the \`tagName\` property 1`] = `
<span
  class=""
/>
`;
